### PR TITLE
Withdraw balance from lock

### PR DESCRIPTION
**Pull request by benwerd**
Refs #529

## Proposed Changes

- Wires up "withdrawal" button to create a transaction that withdraws funds
- Adds withdrawal transaction to redux store so it can be traced
- Displays notice on lock while withdrawal is in progress, both for submission and confirmations
- Does not attempt to withdraw funds if balance on lock is 0

### CODE REVIEW

**Comment by benwerd** [comment]
screen shot 2018-11-14 at 3 36 01 pm

screen shot 2018-11-14 at 3 36 18 pm

File: unlock-app/src/components/creator/CreatorLock.js [hunk 1]
```diff
@@ -37,16 +48,43 @@ export class CreatorLock extends React.Component {
...
+    }
+
     // Some sanitization of strings to display
     let name = lock.name || 'New Lock'
     let outstandingKeys = lock.outstandingKeys || 0
     let lockComponentStatusBlock = (
-      <LockIconBarContainer>
-        <LockIconBar lock={lock} toggleCode={this.toggleEmbedCode} />
-      </LockIconBarContainer>)
+      <StatusBlock>
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLock.js]: What do you think of moving all of the withdraw logic inside `LockIconBar`?

File: unlock-app/src/components/interface/buttons/lock/Withdraw.js [hunk 2]
```diff
@@ -2,10 +2,12 @@ import React from 'react'
 import Svg from '../../svg'
 import { LockButton } from '../Button'
 
-const Withdraw = (props) => (
-  <LockButton title='Withdraw balance' {...props}>
-    <Svg.Withdraw name="Withdraw" />
-  </LockButton>
-)
+export function Withdraw(props) {
```
Comment (julien51) [inline unlock-app/src/components/interface/buttons/lock/Withdraw.js]: huh, why did you move back to a named function here? ES6 syntax is much cleaner?

File: unlock-app/src/pages/dashboard.js [hunk 3]
```diff
@@ -30,13 +30,14 @@ export const Dashboard = ({account, network, locks}) => {
 Dashboard.displayName = 'Dashboard'
 
 Dashboard.propTypes = {
-  account: UnlockPropTypes.account.isRequired,
+  account: UnlockPropTypes.account, // We can't mark this as required because it's undefined when the dashboard is loading
```
Comment (julien51) [inline unlock-app/src/pages/dashboard.js]: Yeah, I think we need to fix that in another PR. I think we actually need to keep it required but not render if it is missing ;)

File: unlock-app/src/propTypes.js [hunk 4]
```diff
@@ -38,7 +38,7 @@ export const mechanism = PropTypes.oneOf(['0', '1', '2', undefined])
 
 export const layout = PropTypes.instanceOf(Function) //PropTypes.instanceOf(React.Component)
 
-export const locks = PropTypes.arrayOf(PropTypes.object)
```
Comment (julien51) [inline unlock-app/src/propTypes.js]: Actually,
Would this `export const locks = PropTypes.arrayOf(lock)` work? ( I think we define the type `lock`)

File: unlock-app/src/stories/creator/CreatorLock.stories.js [hunk 5]
```diff
@@ -32,6 +42,7 @@ storiesOf('CreatorLock', CreatorLock)
   .addDecorator(getStory => <Provider store={store}>{getStory()}</Provider>)
   .add('Deployed', () => {
     const lock = {
+      id: '0xab',
```
Comment (julien51) [inline unlock-app/src/stories/creator/CreatorLock.stories.js]: This is a a giant nit but the id is not the address and starting with `0x` implies an eth address? Would you be ok to use another random id?

File: unlock-app/src/propTypes.js (hunk 4 above)
Comment (benwerd) [reply unlock-app/src/propTypes.js]: It wouldn't right now, because it's not an array - it's an object where the properties are lock IDs. We could shift that in another PR.

Comment (julien51) [reply unlock-app/src/propTypes.js]: ohhh I see! no I think this is right it is better if it is an object whose keys are lock ids. sorry about that!
We can probably do something like that though:

```
PropTypes.objectOf(lock)
```

**Comment by benwerd** [comment]
Moved the withdrawal logic, updated based on feedback. PTAL

**Comment by julien51** [comment]
Wait, did you push your changes? I still see a lot of stuff in `CreatorLock`?

**Comment by benwerd** [comment]
Moved significantly more logic out of `CreatorLock`.

**Comment by julien51** [comment]
Ok you are going to hate me for the next 12 years, but I see a lot of things going on which I believe being in there own PR. Would you mind splitting things? Or would make reviewing much much easier :/

Also, I see you moved a lot of logic down into the lockIconBar (including the status stuff) which is great but then the name is probably not appropriate anymore? I was thinking you could even move the withdraw logic down to the withdraw button (buy that means you'll have to pass the transaction status one level up... Which is I'd say is ok but I'll let you decide).

Also it looks like it's missing tests?

Again, I understand I'm being picky here but I think it's for the best and should make things easier to review and improve in the future!

Thanks

**Comment by benwerd** [comment]
Closing this PR as all component elements have now been submitted separately.
